### PR TITLE
Jax Example

### PR DESCRIPTION
**Pull request by mariaboerner1987**
- create a "from centralized to federated" code example with JAX and linear regression
- add a README for code example
- add pyproject.toml for code example
- add jax documentation
- add jax links to Flower documentation index

### CODE REVIEW

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+        def get_parameters(self) -> Dict:
+            # Return model parameters as a list of NumPy ndarrays,
+            parameter_value = []
+            for _, val in self.params.items():
+                parameter_value.append(np.array(val))
+            return parameter_value
+        
+        def set_parameters(self, parameters: List[np.ndarray]) -> Dict:
+            # Collect model parameters and set new weight values
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:213]:
```suggestion
            # Collect model parameters and update the parameters of the local model
```

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            self.params = self.set_parameters(parameters)
+            loss, num_examples = jax_training.evaluation(self.params,self.grad_fn, self.test_x, self.test_y)
+            print("Evaluation accuracy & loss", loss)
+            return (
+                float(loss),
+                num_examples,
+                {"loss": float(loss)},
+            )
+
+Having defined data loading, model architecture, training, and evaluation we can put everything together and train our :code:`Linear Regression` model.
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:248]: Isn't this the same sentence as above? Should we mention that it's now federated?

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        # Load data
+        train_x, train_y, test_x, test_y = jax_training.load_data()
+        grad_fn = jax.grad(jax_training.loss_fn)
+
+        # Load model (from centralized training) and initialize parameters
+        model_shape = train_x.shape[1:]
+        params = jax_training.load_model(model_shape)
+
+        # Start Flower client
+        client = MNISTClient(params, grad_fn, train_x, train_y, test_x, test_y)
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:264]:
```suggestion
        client = FlowerClient(params, grad_fn, train_x, train_y, test_x, test_y)
```

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+The challenging part is to transform the JAX parameters from :code:`DeviceArray` to :code:`NumPy Arrays` to make it readable for Flower. 
+
+The two :code:`NumPyClient` methods :code:`fit` and :code:`evaluate` make use of the functions :code:`train()` and :code:`evaluate()` previously defined in :code:`jax_trainingt.py`.
+So what we really do here is we tell Flower through our :code:`NumPyClient` subclass which of our already defined functions to call for training and evaluation.
+We included type annotations to give you a better understanding of the data types that get passed around.
+
+.. code-block:: python
+
+
+    class MNISTClient(fl.client.NumPyClient):
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:186]:
```suggestion
    class FlowerClient(fl.client.NumPyClient):
```

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        main()
+
+
+And that's it. You can now open two additional terminal windows and run
+
+.. code-block:: python
+
+    python3 client.py
+
+in each window (make sure that the server is still running before you do so) and see your MXNet project run federated learning across two clients. Congratulations!
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:277]: MXNet?

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+Next Steps
+----------
+
+The full source code for this example: `Jax: From Centralized To Federated (Code) <https://github.com/adap/flower/blob/main/examples/jax_from_centralized_to_federated>`_.
+Our example is of course somewhat over-simplified because both clients load very simple linear regression and the corresponding datset.
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:283]:
```suggestion
Our example is of course somewhat over-simplified because both clients load the same dataset. 
```

File: doc/source/example-jax-from-centralized-to-federated.rst [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    python3 client.py
+
+in each window (make sure that the server is still running before you do so) and see your MXNet project run federated learning across two clients. Congratulations!
+
+Next Steps
+----------
+
+The full source code for this example: `Jax: From Centralized To Federated (Code) <https://github.com/adap/flower/blob/main/examples/jax_from_centralized_to_federated>`_.
+Our example is of course somewhat over-simplified because both clients load very simple linear regression and the corresponding datset. 
+You're now prepared to explore this topic further. How about using a CNN or using a different dataset? How about adding more clients?
```
Comment (danieljanes) [inline doc/source/example-jax-from-centralized-to-federated.rst:284]:
```suggestion
You're now prepared to explore this topic further. How about using a more sophisticated model or using a different dataset? How about adding more clients?
```

File: examples/jax_from_centralized_to_federated/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
+# JAX: From Centralized To Federated
+
+This example demonstrates how an already existing centralized JAX-based machine learning project can be federated with Flower.
+
+This introductory example for Flower uses JAX, but you're not required to be a JAX expert to run the example. The example will help you to understand how Flower can be used to build federated learning use cases based on an existing MXNet project.
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/README.md:5]: MXNet?

File: examples/jax_from_centralized_to_federated/README.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+```shell
+python3 -c "import flwr"
+```
+
+If you don't see any errors you're good to go!
+
+## Run JAX Federated
+
+This MXNet example is based on the [Linear Regression with JAX](https://coax.readthedocs.io/en/latest/examples/linear_regression/jax.html) tutorial and uses a sklearn dataset (generating a random dataset for a regression pronlem). Feel free to consult the tutorial if you want to get a better understanding of JAX. If you play around with the dataset, please keep in mind that the data samples are generated randomly depending on the settings being done while calling the dataset function. Please checkout out the [sklearn tutorial for further information](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html). The file `jax_training.py` contains all the steps that are described in the tutorial. It loads the train and test dataset and a linear regression model, trains the model with the training set, and evaluates the trained model on the test set.
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/README.md:42]:
```suggestion
This MXNet example is based on the [Linear Regression with JAX](https://coax.readthedocs.io/en/latest/examples/linear_regression/jax.html) tutorial and uses a sklearn dataset (generating a random dataset for a regression pronlem). Feel free to consult the tutorial if you want to get a better understanding of JAX. If you play around with the dataset, please keep in mind that the data samples are generated randomly depending on the settings being done while calling the dataset function. Please checkout out the [scikit-learn tutorial for further information](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html). The file `jax_training.py` contains all the steps that are described in the tutorial. It loads the train and test dataset and a linear regression model, trains the model with the training set, and evaluates the trained model on the test set.
```

File: examples/jax_from_centralized_to_federated/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+python3 client.py
+```
+
+Start client 2 in the second terminal:
+
+```shell
+python3 client.py
+```
+
+You are now training a JAX-based linear regression model, federated across two clients. The setup is of course simplified since both clients hold a similar dataset, but you can now continue with your own explorations. How about changing from a linear regression to a CNN? How about adding more clients?
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/README.md:66]:
```suggestion
You are now training a JAX-based linear regression model, federated across two clients. The setup is of course simplified since both clients hold a similar dataset, but you can now continue with your own explorations. How about changing from a linear regression to a more sophisticated model? How about adding more clients?
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import flwr as fl
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+import jax_training
+
+
+# Flower Client
+class MNISTClient(fl.client.NumPyClient):
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:16]:
```suggestion
class FlowerClient(fl.client.NumPyClient):
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    """Flower client implementing using linear regression and JAX"""
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:17]:
```suggestion
    """Flower client implementing using linear regression using JAX"""
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    ) -> None:
+        self.params= params
+        self.grad_fn = grad_fn
+        self.train_x = train_x
+        self. train_y = train_y
+        self.test_x = test_x
+        self.test_y = test_y
+
+    def get_parameters(self):
+        # Return model parameters as a list of NumPy ndarrays,
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:36]:
```suggestion
        # Return model parameters as a list of NumPy ndarrays
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        parameter_value = []
+        for _, val in self.params.items():
+            parameter_value.append(np.array(val))
+        return parameter_value
+    
+    def set_parameters(self, parameters: List[np.ndarray]) -> None:
+        # Collect model parameters and set new weight values
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:43]:
```suggestion
        # Collect model parameters and update the parameters of the local model
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        grad_fn: Callable,
+        train_x: List[np.ndarray],
+        train_y: List[np.ndarray],
+        test_x: List[np.ndarray],
+        test_y: List[np.ndarray],
+    ) -> None:
+        self.params= params
+        self.grad_fn = grad_fn
+        self.train_x = train_x
+        self. train_y = train_y
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:31]: Does this work?

File: examples/jax_from_centralized_to_federated/client.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        print("Evaluation accuracy & loss", loss)
+        return (
+            float(loss),
+            num_examples,
+            {"loss": float(loss)},
+        )
+
+
+def main() -> None:
+    """Load data, start MNISTClient."""
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:80]:
```suggestion
    """Load data, start FlowerClient."""
```

File: examples/jax_from_centralized_to_federated/client.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    # Load data
+    train_x, train_y, test_x, test_y = jax_training.load_data()
+    grad_fn = jax.grad(jax_training.loss_fn)
+
+    # Load model (from centralized training) and initialize parameters
+    model_shape = train_x.shape[1:]
+    params = jax_training.load_model(model_shape)
+
+    # Start Flower client
+    client = MNISTClient(params, grad_fn, train_x, train_y, test_x, test_y)
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/client.py:91]:
```suggestion
    client = FlowerClient(params, grad_fn, train_x, train_y, test_x, test_y)
```

File: examples/jax_from_centralized_to_federated/pyproject.toml [hunk 18]
```diff
@@ -0,0 +1,19 @@
+[tool.poetry]
+name = "jax_example"
+version = "0.1.0"
+description = "JAX example running a linear regression model federated"
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/pyproject.toml:4]:
```suggestion
description = "JAX example training a linear regression model with federated learning"
```

File: examples/jax_from_centralized_to_federated/pyproject.toml [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+authors = ["The Flower Authors <enquiries@flower.dev>"]
+
+[tool.poetry.dependencies]
+python = "^3.8"
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/pyproject.toml:8]:
```suggestion
python = "^3.7"
```

File: examples/jax_from_centralized_to_federated/pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+jax = "^0.2.25"
+jaxlib = "^0.1.74"
+sklearn = "^1.0.1"
+flwr = "^0.17.0"
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/pyproject.toml:12]:
```suggestion
flwr = "^0.18.0"
```

File: examples/jax_from_centralized_to_federated/pyproject.toml [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+numpy = "^1.21.4"
+
+[tool.poetry.dev-dependencies]
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/pyproject.toml:15]: This can be removed

File: examples/jax_from_centralized_to_federated/jax_training.py [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        loss = loss_fn(params,X, y)
+        if epochs % 10 == 0:
+            print(f'For Epoch {epochs} loss {loss}')
+    return params, loss, num_examples
+
+def evaluation(params, grad_fn, X_test, y_test) -> Tuple[float, int]:
+    num_examples = X_test.shape[0]
+    err_test = loss_fn(params, X_test, y_test)
+    loss_test = jnp.mean(jnp.square(err_test))
+    #print(f'Test loss {loss_test}')
```
Comment (danieljanes) [inline examples/jax_from_centralized_to_federated/jax_training.py:53]:
```suggestion
    # print(f'Test loss {loss_test}')
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
Code example and documentation look good, I've added a few suggestions. One open question: should we call this "from centralized to federated" or "quickstart JAX"?

**Comment by mariaboerner1987** [comment]
> Code example and documentation look good, I've added a few suggestions. One open question: should we call this "from centralized to federated" or "quickstart JAX"?

From centralized to federated makes more sense to me.

**Review by danieljanes — APPROVED** [review]
